src/core/SampleBuffer.cpp

```diff
@@ -957,12 +925,18 @@ void SampleBuffer::visualize( QPainter & _p, const QRect & _dr,
 	const int xb = _dr.x();
 	const int first = focus_on_range ? _from_frame : 0;
 	const int last = focus_on_range ? _to_frame : m_frames;
-	for( int frame = first; frame < last; frame += fpp )
-	{
-		l[n] = QPointF( xb + ( (frame - first) * double( w ) / nb_frames ),
-			( yb - ( m_data[frame][0] * y_space * m_amplification ) ) );
-		r[n] = QPointF( xb + ( (frame - first) * double( w ) / nb_frames ),
-			( yb - ( m_data[frame][1] * y_space * m_amplification ) ) );
+	
+	const bool isReversed = m_reversed;
+
+	auto startFrame = isReversed ? last : first;
+	auto count = (last - first) / fpp + 1;
```

Comment:
I'm not sure if maybe this is a scenario that will never happen, but before the `for` loop ran starting on `first` while the current frame was below `last`. So if first was equal to last, the loop would never run.

With the current logic, if last is equal to first, count will be 1, and with `i` beginning at 0 it will run the loop once. I think I understood the `+1` on `count` (so if `last - first` is smaller than `fpp` it will still draw the frames), but is there a chance that `last` will be equal to `first`?

Reply:
Oh, I thought it was `[first, last]`. If we need `[first, last)` or `(first, last]`, it should be `(last - first + fpp - 1) / fpp`.

Reply:
I don't know, to be honest the previous loop could be the wrong one, I just compared both. If `first` is equal to `last` do we have a buffer with length of 0 or 1? From the looks of it it seems that last is non-inclusive, since when `focus_on_range == false` it can be set to `m_frames`, which I think is the total number of frames. So `last` would only be equal to `first` if there are no frames. We just need to make sure and use your suggestion if it's non-inclusive,

Reply:
From what I've been reading, `m_frames` is the length of the sample in frames. Even when a sample fails to load, a 1-frame sample is created (`m_frames = 1`). If `first` starts at 0 and `last` is equal to the total number of frames then `last` is non-inclusive. Apparently `last` being equal to `first` would mean we have a sample buffer of length 0, and this doesn't seem to happen.

The weird part is that when `focus_on_range` is true, `last` doesn't seem to be a length anymore, but instead a frame index (`_to_frame`). If we are using `last` as an index, then maybe we should have `const int last = focus_on_range ? _to_frame : m_frames - 1;` right?

I'm still reading some of the code to make sure, if someone could help me check this out. Sorry if I'm making a bit of a confusion here, I just want to make sure we have a proper loop.